File: crates/billing/src/plans.rs
use serde::{Deserialize, Serialize};

/// Subscription plan types
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Plan {
    Free,
    Pro,
    Team,
    Enterprise,
}

impl Default for Plan {
    fn default() -> Self {
        Self::Free
    }
}

impl std::fmt::Display for Plan {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Plan::Free => write!(f, "free"),
            Plan::Pro => write!(f, "pro"),
            Plan::Team => write!(f, "team"),
            Plan::Enterprise => write!(f, "enterprise"),
        }
    }
}

impl Plan {
    pub fn from_stripe_price_id(price_id: &str) -> Option<Self> {
        // These should match your Stripe price IDs
        // In production, load these from environment or config
        match price_id {
            id if id.starts_with("price_pro_") => Some(Plan::Pro),
            id if id.starts_with("price_team_") => Some(Plan::Team),
            id if id.starts_with("price_enterprise_") => Some(Plan::Enterprise),
            _ => None,
        }
    }

    pub fn limits(&self) -> PlanLimits {
        PLANS.iter()
            .find(|p| p.plan == *self)
            .map(|p| p.limits.clone())
            .unwrap_or_default()
    }
}

/// Plan limits and quotas
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanLimits {
    /// Maximum number of servers
    pub max_servers: u32,
    /// Maximum number of deployments per month
    pub max_deployments_per_month: u32,
    /// Maximum number of API requests per month
    pub max_requests_per_month: u64,
    /// Maximum number of team members (for Team/Enterprise)
    pub max_team_members: u32,
    /// Log retention in days
    pub log_retention_days: u32,
    /// Custom domain support
    pub custom_domains: bool,
    /// Priority support
    pub priority_support: bool,
    /// SSO support
    pub sso_enabled: bool,
}

impl Default for PlanLimits {
    fn default() -> Self {
        // Free tier limits
        Self {
            max_servers: 3,
            max_deployments_per_month: 50,
            max_requests_per_month: 10_000,
            max_team_members: 1,
            log_retention_days: 7,
            custom_domains: false,
            priority_support: false,
            sso_enabled: false,
        }
    }
}

/// Plan definition with pricing info
#[derive(Debug, Clone, Serialize)]
pub struct PlanDefinition {
    pub plan: Plan,
    pub name: &'static str,
    pub description: &'static str,
    pub price_monthly_usd: u32,
    pub price_yearly_usd: u32,
    #[serde(skip)]
    pub stripe_price_id_monthly: Option<&'static str>,
    #[serde(skip)]
    pub stripe_price_id_yearly: Option<&'static str>,
    pub limits: PlanLimits,
    #[serde(serialize_with = "serialize_features")]
    pub features: &'static [&'static str],
}

fn serialize_features<S>(features: &&'static [&'static str], serializer: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    use serde::ser::SerializeSeq;
    let mut seq = serializer.serialize_seq(Some(features.len()))?;
    for feature in *features {
        seq.serialize_element(feature)?;
    }
    seq.end()
}

/// All available plans
pub static PLANS: &[PlanDefinition] = &[
    PlanDefinition {
        plan: Plan::Free,
        name: "Free",
        description: "Perfect for getting started",
        price_monthly_usd: 0,
        price_yearly_usd: 0,
        stripe_price_id_monthly: None,
        stripe_price_id_yearly: None,
        limits: PlanLimits {
            max_servers: 3,
            max_deployments_per_month: 50,
            max_requests_per_month: 10_000,
            max_team_members: 1,
            log_retention_days: 7,
            custom_domains: false,
            priority_support: false,
            sso_enabled: false,
        },
        features: &[
            "3 MCP servers",
            "50 deployments/month",
            "10K API requests/month",
            "7-day log retention",
            "Community support",
        ],
    },
    PlanDefinition {
        plan: Plan::Pro,
        name: "Pro",
        description: "For professional developers",
        price_monthly_usd: 29,
        price_yearly_usd: 290,
        stripe_price_id_monthly: None, // Set via STRIPE_PRICE_PRO_MONTHLY env
        stripe_price_id_yearly: None,  // Set via STRIPE_PRICE_PRO_YEARLY env
        limits: PlanLimits {
            max_servers: 20,
            max_deployments_per_month: 500,
            max_requests_per_month: 500_000,
            max_team_members: 1,
            log_retention_days: 30,
            custom_domains: true,
            priority_support: false,
            sso_enabled: false,
        },
        features: &[
            "20 MCP servers",
            "500 deployments/month",
            "500K API requests/month",
            "30-day log retention",
            "Custom domains",
            "Email support",
        ],
    },
    PlanDefinition {
        plan: Plan::Team,
        name: "Team",
        description: "For growing teams",
        price_monthly_usd: 99,
        price_yearly_usd: 990,
        stripe_price_id_monthly: None, // Set via STRIPE_PRICE_TEAM_MONTHLY env
        stripe_price_id_yearly: None,  // Set via STRIPE_PRICE_TEAM_YEARLY env
        limits: PlanLimits {
            max_servers: 100,
            max_deployments_per_month: 2000,
            max_requests_per_month: 5_000_000,
            max_team_members: 10,
            log_retention_days: 90,
            custom_domains: true,
            priority_support: true,
            sso_enabled: false,
        },
        features: &[
            "100 MCP servers",
            "2000 deployments/month",
            "5M API requests/month",
            "90-day log retention",
            "Custom domains",
            "Up to 10 team members",
            "Priority support",
        ],
    },
    PlanDefinition {
        plan: Plan::Enterprise,
        name: "Enterprise",
        description: "For large organizations",
        price_monthly_usd: 499,
        price_yearly_usd: 4990,
        stripe_price_id_monthly: None, // Set via STRIPE_PRICE_ENTERPRISE_MONTHLY env
        stripe_price_id_yearly: None,  // Set via STRIPE_PRICE_ENTERPRISE_YEARLY env
        limits: PlanLimits {
            max_servers: u32::MAX,
            max_deployments_per_month: u32::MAX,
            max_requests_per_month: u64::MAX,
            max_team_members: u32::MAX,
            log_retention_days: 365,
            custom_domains: true,
            priority_support: true,
            sso_enabled: true,
        },
        features: &[
            "Unlimited MCP servers",
            "Unlimited deployments",
            "Unlimited API requests",
            "1-year log retention",
            "Custom domains",
            "Unlimited team members",
            "SSO/SAML",
            "Dedicated support",
            "SLA",
        ],
    },
];

/// Get plan definition by plan type
pub fn get_plan_definition(plan: Plan) -> Option<&'static PlanDefinition> {
    PLANS.iter().find(|p| p.plan == plan)
}

/// Get plan by Stripe price ID
pub fn get_plan_by_price_id(price_id: &str) -> Option<Plan> {
    // Check environment variables for price IDs
    if let Ok(pro_monthly) = std::env::var("STRIPE_PRICE_PRO_MONTHLY") {
        if price_id == pro_monthly {
            return Some(Plan::Pro);
        }
    }
    if let Ok(pro_yearly) = std::env::var("STRIPE_PRICE_PRO_YEARLY") {
        if price_id == pro_yearly {
            return Some(Plan::Pro);
        }
    }
    if let Ok(team_monthly) = std::env::var("STRIPE_PRICE_TEAM_MONTHLY") {
        if price_id == team_monthly {
            return Some(Plan::Team);
        }
    }
    if let Ok(team_yearly) = std::env::var("STRIPE_PRICE_TEAM_YEARLY") {
        if price_id == team_yearly {
            return Some(Plan::Team);
        }
    }
    if let Ok(enterprise_monthly) = std::env::var("STRIPE_PRICE_ENTERPRISE_MONTHLY") {
        if price_id == enterprise_monthly {
            return Some(Plan::Enterprise);
        }
    }
    if let Ok(enterprise_yearly) = std::env::var("STRIPE_PRICE_ENTERPRISE_YEARLY") {
        if price_id == enterprise_yearly {
            return Some(Plan::Enterprise);
        }
    }

    None
}

File: apps/web/src/lib/plans.ts
// Shared plan definitions - must match backend crates/billing/src/plans.rs

export interface PlanLimits {
  max_servers: number;
  max_deployments_per_month: number;
  max_requests_per_month: number;
  max_team_members: number;
  log_retention_days: number;
  custom_domains: boolean;
  priority_support: boolean;
  sso_enabled: boolean;
}

export interface PlanDefinition {
  plan: 'free' | 'pro' | 'team' | 'enterprise';
  name: string;
  description: string;
  price_monthly_jpy: number;
  price_yearly_jpy: number;
  limits: PlanLimits;
  features: string[];
}

// 1 USD = 150 JPY (approximate, rounded for clean pricing)
// Backend: Free $0, Pro $29, Team $99, Enterprise $499
export const PLANS: PlanDefinition[] = [
  {
    plan: 'free',
    name: 'Free',
    description: '個人開発や検証に最適',
    price_monthly_jpy: 0,
    price_yearly_jpy: 0,
    limits: {
      max_servers: 3,
      max_deployments_per_month: 50,
      max_requests_per_month: 10_000,
      max_team_members: 1,
      log_retention_days: 7,
      custom_domains: false,
      priority_support: false,
      sso_enabled: false,
    },
    features: [
      'サーバー3つまで',
      'デプロイ50回/月',
      '月間10,000リクエスト',
      'ログ7日間保持',
      'コミュニティサポート',
    ],
  },
  {
    plan: 'pro',
    name: 'Pro',
    description: '本番運用に必要な全機能',
    price_monthly_jpy: 2980,
    price_yearly_jpy: 29800, // ~17% off
    limits: {
      max_servers: 20,
      max_deployments_per_month: 500,
      max_requests_per_month: 500_000,
      max_team_members: 1,
      log_retention_days: 30,
      custom_domains: true,
      priority_support: false,
      sso_enabled: false,
    },
    features: [
      'サーバー20個まで',
      'デプロイ500回/月',
      '月間500,000リクエスト',
      'ログ30日間保持',
      'カスタムドメイン',
      'メールサポート',
    ],
  },
  {
    plan: 'team',
    name: 'Team',
    description: 'チームでの本格運用に',
    price_monthly_jpy: 9800,
    price_yearly_jpy: 98000, // ~17% off
    limits: {
      max_servers: 100,
      max_deployments_per_month: 2000,
      max_requests_per_month: 5_000_000,
      max_team_members: 10,
      log_retention_days: 90,
      custom_domains: true,
      priority_support: true,
      sso_enabled: false,
    },
    features: [
      'サーバー100個まで',
      'デプロイ2000回/月',
      '月間5,000,000リクエスト',
      'ログ90日間保持',
      'カスタムドメイン',
      'チームメンバー10人まで',
      '優先サポート',
    ],
  },
  {
    plan: 'enterprise',
    name: 'Enterprise',
    description: '大規模組織向け',
    price_monthly_jpy: 49800,
    price_yearly_jpy: 498000, // ~17% off
    limits: {
      max_servers: Infinity,
      max_deployments_per_month: Infinity,
      max_requests_per_month: Infinity,
      max_team_members: Infinity,
      log_retention_days: 365,
      custom_domains: true,
      priority_support: true,
      sso_enabled: true,
    },
    features: [
      'サーバー無制限',
      'デプロイ無制限',
      'リクエスト無制限',
      'ログ1年間保持',
      'カスタムドメイン',
      'チームメンバー無制限',
      'SSO/SAML',
      '専任サポート',
      'SLA保証',
    ],
  },
];

export function getPlan(planId: string): PlanDefinition | undefined {
  return PLANS.find(p => p.plan === planId);
}

export function formatPrice(price: number): string {
  return `¥${price.toLocaleString()}`;
}
